blockchain/utxocache.go

```diff
@@ -0,0 +1,1085 @@
+	mtx sync.Mutex
+
+	// cachedEntries keeps the internal cache of the utxo state.  The tfModified
+	// flag indicates that the state of the entry (potentially) deviates from the
+	// state in the database.  Explicit nil values in the map are used to
+	// indicate that the database does not contain the entry.
+	cachedEntries    mapSlice
+	totalEntryMemory uint64 // Total memory usage in bytes.
+	lastFlushHash    chainhash.Hash
+}
+
+// newUtxoCache initiates a new utxo cache instance with its memory usage limited
+// to the given maximum.
+func newUtxoCache(db database.DB, maxTotalMemoryUsage uint64) *utxoCache {
+	// While the entry isn't included in the map size, add the average size to the
+	// bucket size so we get some leftover space for entries to take up.
+	numMaxElements := CalculateMinEntries(int(maxTotalMemoryUsage), bucketSize+avgEntrySize)
+	numMaxElements -= 1
+
+	log.Infof("Pre-alloacting for %d MiB: ", maxTotalMemoryUsage/(1024*1024)+1)
+
+	m := make(map[wire.OutPoint]*UtxoEntry, numMaxElements)
+
+	return &utxoCache{
+		db:                  db,
+		maxTotalMemoryUsage: maxTotalMemoryUsage,
+		cachedEntries: mapSlice{
+			maps:                []map[wire.OutPoint]*UtxoEntry{m},
+			maxEntries:          []int{numMaxElements},
+			maxTotalMemoryUsage: maxTotalMemoryUsage,
+		},
+	}
+}
+
+// totalMemoryUsage returns the total memory usage in bytes of the UTXO cache.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) totalMemoryUsage() uint64 {
+	// Total memory is the map size + the size that the utxo entries are
+	// taking up.
+	size := uint64(s.cachedEntries.size())
+	size += s.totalEntryMemory
+
+	return size
+}
+
+// TotalMemoryUsage returns the total memory usage in bytes of the UTXO cache.
+//
+// This method is safe for concurrent access.
+func (s *utxoCache) TotalMemoryUsage() uint64 {
+	s.mtx.Lock()
+	tmu := s.totalMemoryUsage()
+	s.mtx.Unlock()
+	return tmu
+}
+
+// fetchAndCacheEntries attempts to fetch a series of entries from the
+// database.  In any aren't found, then nil is returned. All entries found are
+// cached in the process.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) fetchAndCacheEntries(ops []wire.OutPoint) ([]*UtxoEntry, error) {
+	entries := make([]*UtxoEntry, len(ops))
+	err := s.db.View(func(dbTx database.Tx) error {
+		utxoBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+
+		for i, op := range ops {
+			entry, err := dbFetchUtxoEntry(dbTx, utxoBucket, op)
+			if err != nil {
+				return err
+			}
+
+			entries[i] = entry
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Add each of the entries to the UTXO cache and update their memory
+	// usage.
+	//
+	// NOTE: When the fetched entry is nil, it is still added to the cache
+	// as a miss; this prevents future lookups to perform the same database
+	// fetch.
+	for i, entry := range entries {
+		s.totalEntryMemory += entry.memoryUsage()
```

Comment:
Won't this be wrong if the entry is already in the cache?

Reply:
Could consider moving the updates to `totalEntryMemory` inside the `put/delete` methods.

Reply:
> Won't this be wrong if the entry is already in the cache?

`fetchAndCacheEntries()` is only called by `getEntry()` and `getEntries()` after they have checked in the cache. So all the outpoints they're fetching are guaranteed to not be in the cache.

I guess this is confusing. Maybe move the function inside `getEntries()` and have `getEntry()` just call `genEntries()` or just get rid of `getEntries()` entirely.